src/main/kotlin/com/ajaxproject/warehouse/beanpostprocessor/RequestLimitAnnotationBeanPostProcessor.kt

```diff
@@ -0,0 +1,83 @@
+package com.ajaxproject.warehouse.beanpostprocessor
+
+import com.ajaxproject.warehouse.annotation.RequestLimit
+import com.ajaxproject.warehouse.exception.MethodRequestLimitExceededException
+import org.springframework.beans.factory.config.BeanPostProcessor
+import org.springframework.cglib.proxy.InvocationHandler
+import org.springframework.cglib.proxy.Proxy
+import org.springframework.stereotype.Component
+import org.springframework.stereotype.Service
+import java.lang.reflect.Method
+import java.util.concurrent.atomic.AtomicInteger
+
+@Component
+class RequestLimitAnnotationBeanPostProcessor : BeanPostProcessor {
+
+    val annotatedBeansMap: HashMap<String, Class<Any>> = HashMap()
+
+    override fun postProcessBeforeInitialization(bean: Any, beanName: String): Any? {
+        val beanClass = bean.javaClass
+        if (beanClass.isAnnotationPresent(Service::class.java)) {   // TODO ask about interfaces and other
+            if (beanClass.methods.any { it.isAnnotationPresent(RequestLimit::class.java) }) {
+                annotatedBeansMap[beanName] = beanClass
+            }
+        }
+        return bean
+    }
+
+    override fun postProcessAfterInitialization(bean: Any, beanName: String): Any? {
+        val beanClass = annotatedBeansMap[beanName] ?: return bean
+
+        return Proxy.newProxyInstance(
+            beanClass.classLoader,
+            beanClass.interfaces,
+            RequestLimitInvocationHandler(bean, beanClass)
+        )
+    }
+}
+
+class RequestLimitInvocationHandler(
+    private val bean: Any,
+    private val beanClass: Class<Any>
+) : InvocationHandler {
+
+    private val methodRequests: HashMap<String, AtomicInteger> = HashMap()
```

Comment:
```suggestion
    private val methodRequests: MutableMap<String, AtomicInteger> = mutableMapOf()
```

Reply:
by the way, you should use concurrent hash map here instead of thread unsafe collection